src/backend/common/util.cpp

```diff
@@ -84,3 +91,86 @@ void saveKernel(const std::string& funcName, const std::string& jit_ker,
     }
     fclose(f);
 }
+
+#if !defined(OS_WIN)
+string getHomeDirectory() {
+    string home = getEnvVar("XDG_CACHE_HOME");
+    if (!home.empty()) return home;
+
+    home = getEnvVar("HOME");
+    if (!home.empty()) return home;
+
+    home = getpwuid(getuid())->pw_dir;
+}
+#else
+string getTemporaryDirectory() {
+    DWORD bufSize = 261;  // limit according to GetTempPath documentation
+    string retVal;
+    retVal.resize(bufSize);
+    bufSize = GetTempPathA(bufSize, &retVal[0]);
+    retVal.resize(bufSize);
+    return retVal;
+}
+#endif
+
+bool folderExists(const string& path) {
+#if !defined(OS_WIN)
+    struct stat status;
+    return stat(path.c_str(), &status) == 0 && (status.st_mode & S_IFDIR) != 0;
+#else
+    struct _stat status;
+    return _stat(path.c_str(), &status) == 0 && (status.st_mode & S_IFDIR) != 0;
+#endif
+}
+
+bool createFolder(const string& path) {
+#if !defined(OS_WIN)
+    return mkdir(path.c_str(), 0777) == 0;
```

Comment:
I am not too versed in these permission strings. Isn't the "execute" bit required for simple things like "ls"? Also I think we can't make this directory only owned by the current user, as if the folder is somewhere on the root (like "/var/tmp/arrayfire/") this could potentially be accessed by multiple users.

Reply:
the execute bit is needed on binary executable but since we will only be writing once and reading multiple times 644 would be a conservative option. Other users and groups can still read the file but they don't necessarily need to write there.

Reply:
Note this is the permission for the directory, not for the files that get written in it. For directories, the execute bit must be set at least for the current user, [otherwise you cannot create / remove files](https://unix.stackexchange.com/questions/21251/execute-vs-read-bit-how-do-directory-permissions-in-linux-work). Other users might need write access to that directory if they want to cache their own kernels there too (if that directory is not in the users's home, that is).

Reply:
Ahh. my mistake. never mind then.

Reply:
In my home directory on my Linux Mint machine, most directories have permission 0755, but ``.local`` is 0700. I'd go with 0755 if the directory is in the user's home folder, and 0777 otherwise.